### PR TITLE
Adding some compatibility, with room for further compat in the future.

### LINKED ISSUE #219: Miner destroys broken Tinker Construct tools when they run out of durability
### Minecolonies version
0.4.1080
### Expected behavior
Tinker construct tools do not get destroyed when they run out of durability, they simply become "broken" and can be repaired
### Actual behaviour
However when given a tinker construct pickaxe to the miner, when the durability rans out, it gets destroyed,
### Steps to reproduce the problem
give tinker construct pickaxe to the miner when they ask for it, and wait until it rans out of durability.

### PR DESCRIPTION
**Pull request by Dadadah**
Closes #219

# Changes proposed in this pull request:
- Add a compatibility package
- Check to see if a tinker's tool is broken before returning the mining level

Review please @Minecolonies/maintainers

### CODE REVIEW

File: src/main/java/com/minecolonies/util/Utils.java [hunk 1]
```diff
@@ -376,6 +377,10 @@ public static int getMiningLevel(@Nullable ItemStack stack, @Nullable String too
         {
             return -1;
         }
+        if (Compatibility.checkMiningCompatibility(stack, tool))
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/util/Utils.java]: I'd suggest a different naming, more like `checkTinkersToolBroken`

File: src/main/java/com/compatibility/Compatibility.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.compatibility;
+
+import com.compatibility.tinkers.ToolBrokenCheck;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import org.jetbrains.annotations.Nullable;
+
+public class Compatibility
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1118)
MINOR Document this public class. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: src/main/java/com/compatibility/Compatibility.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public static boolean checkMiningCompatibility(@Nullable ItemStack stack, @Nullable String tool)
+    {
+        if (ToolBrokenCheck.checkTinkersBroken(stack, tool))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MINOR Replace this if-then-else statement by a single return statement. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1126)

File: src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.compatibility.tinkers;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import org.jetbrains.annotations.Nullable;
+
+public class ToolBrokenCheck
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1118)
MINOR Document this public class. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{
+    private static final String STATS = "Stats";
+    private static final String BROKEN = "Broken";
+    public static boolean checkTinkersBroken(@Nullable ItemStack stack, @Nullable String tool)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java]: MAJOR Remove this unused method parameter "tool". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1172)
MINOR Document this public method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 11 issues
* MAJOR 3 major
* MINOR 8 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/3f6874d7b44bf39c6327bf955c291d13640013ad/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/3f6874d7b44bf39c6327bf955c291d13640013ad/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

File: src/main/java/com/minecolonies/util/Utils.java (hunk 1 above)
Comment (Kostronor) [inline src/main/java/com/minecolonies/util/Utils.java]: Or after seeing the other code, more like `checkModCompatibility`

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 10 issues
* MAJOR 3 major
* MINOR 7 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/3f6874d7b44bf39c6327bf955c291d13640013ad/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/3f6874d7b44bf39c6327bf955c291d13640013ad/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

File: src/main/java/com/compatibility/Compatibility.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package com.compatibility;
+
+import com.compatibility.tinkers.ToolBrokenCheck;
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.Nullable;
+
+/**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/compatibility/Compatibility.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    public static boolean checkMiningCompatibility(@Nullable ItemStack stack, @Nullable String tool)
+    {
+        if (ToolBrokenCheck.checkTinkersBroken(stack, tool))
+        {
+            return true;
+        }
+        return false;
+    }
+
+    public Compatibility()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MAJOR Hide this public constructor. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1118)
MINOR Move this constructor to comply with Java Code Conventions. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1213)

File: src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
+package com.compatibility.tinkers;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import org.jetbrains.annotations.Nullable;
+
+/**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 9 issues
* MAJOR 3 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/8f30348fea545d4bb386688328bd63127b16633b/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/8f30348fea545d4bb386688328bd63127b16633b/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

File: src/main/java/com/compatibility/Compatibility.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.compatibility.tinkers.ToolBrokenCheck;
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.Nullable;
+
+/**
+  * This class is to store the methods that call the methods
+  * to check for miscellaneous compatibility problems
+  */
+public class Compatibility
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MAJOR Make this class "final" or add a public constructor. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2974)

File: src/main/java/com/compatibility/Compatibility.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+      * This method checks to see if STACK is able to mine anything.
+      * It goes through all compatibility checks.
+      *
+      * @param stack the item in question.
+      * @param tool the name of the tool.
+      * @return boolean whether the stack can mine or not.
+      */
+    public static boolean checkMiningCompatibility(@Nullable ItemStack stack, @Nullable String tool)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/Compatibility.java]: MAJOR Remove this unused method parameter "tool". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1172)

File: src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  * This class is to store a check to see
+  * if a tinker's tool is broken
+  */
+public class ToolBrokenCheck
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/tinkers/ToolBrokenCheck.java]: MAJOR Make this class "final" or add a public constructor. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2974)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 7 issues
* MAJOR 3 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/e888cd526728fa5449ee5f97e6d1cc1e0b2f4497/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/e888cd526728fa5449ee5f97e6d1cc1e0b2f4497/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 1 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/f67ce960542ea13ad9215096dba8ccf755996b58/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/f67ce960542ea13ad9215096dba8ccf755996b58/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

File: src/main/java/com/compatibility/Compatibility.java (hunk 10 above)
Comment (Dadadah) [inline src/main/java/com/compatibility/Compatibility.java]: Tool is left here for future compatibility

File: src/main/java/com/minecolonies/util/Utils.java (hunk 1 above)
Comment (Dadadah) [inline src/main/java/com/minecolonies/util/Utils.java]: Maybe checkMiningModCompatibility?

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/2a9be162f69a1d8621483a350c6a55200231f893/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/2a9be162f69a1d8621483a350c6a55200231f893/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [compatibility](https://github.com/Minecolonies/minecolonies/blob/455f3f30ca73991e82febc809d909787db40e739/src/main/java/com/compatibility): Add a 'package-info.java' file to document the 'compatibility' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)
1. MINOR [tinkers](https://github.com/Minecolonies/minecolonies/blob/455f3f30ca73991e82febc809d909787db40e739/src/main/java/com/compatibility/tinkers): Add a 'package-info.java' file to document the 'tinkers' package [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

File: src/main/java/com/compatibility/Compatibility.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+      * This method checks to see if STACK is able to mine anything.
+      * It goes through all compatibility checks.
+      *
+      * @param stack the item in question.
+      * @param tool the name of the tool.
+      * @return boolean whether the stack can mine or not.
+      */
+    public static boolean getMiningLevelCompatibility(@Nullable ItemStack stack, @Nullable String tool)
```
Comment (Dadadah) [inline src/main/java/com/compatibility/Compatibility.java]: Tool is here for future compatibility

File: src/main/java/com/compatibility/tinkers/package-info.java [hunk 13]
```diff
@@ -0,0 +1,4 @@
+/**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/compatibility/tinkers/package-info.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by Kostronor** [comment]
LGTM @cltnschlosser @Raycoms

**Comment by cltnschlosser** [comment]
If this is code you've written then it should be under the minecolonies package. If not we should give credit to the original authors.

**Comment by Dadadah** [comment]
This is code I have written, however krostronor was the one who told me to put it in this location.

**Comment by Kostronor** [comment]
@cltnschlosser I really thing we should extract mod compatibility as much as possible. And we already have 3 packages, why not use it?

**Comment by cltnschlosser** [comment]
The this should be created as something that could be standalone (used in other mods). It's fine now, but in the future if we wanted to use minecolonies code inside compatibility code, that would be bad practice.

**Comment by Kostronor** [comment]
We would have to merge it then. At the moment it is planned that minecolonies depends on the compatibility package. Once we have an api, I'll try to convert it so that it's the other way around...

**Pull request merged**